<img src="Logo.png" align="right" width="150" />

##
***
## (950457) Medidas Electrónicas I
#####

Profesor: Ing. Emiliano Marinsek	<<emarinsek@frba.utn.edu.ar>> <br>
Ayudante: Ing. Juan Manuel Perdomo	<<jmperdomo@frba.utn.edu.ar>>

***
####
### <u> Ejercicios de incertidumbre</u>
##### *Introducción al cálculo de incertidumbre*
####

Por David Moharos

***

<img src="Consigna.png" style="margin-left:auto; margin-right:auto"/>

In [214]:
#%% Librerías y módulos de Python a utilizar

import numpy as np
from IPython.display import display, Markdown, HTML
import pandas as pd

In [175]:
#%% Métodos auxiliares

#-- Ejercicio 1 --
def calculo_em (R1, R2):
    return ( R2 / (R2 - R1) )

#-- Ejercicio 2 --
def calculo_deltaV (V1, V2):
    return ( V1 - V2 )

#-- Ejercicio 3 --
def calculo_Vout (Voff, a_s, g, theta):
    return ( Voff + a_s / g * g * np.sin(theta) )

#-- Ejercicio 4 --
def calculo_R1 (V1, I1):
    return ( V1 / I1 )
    
#-- Ejercicio 5--
def calculo_V0 (H1, H0, KH, S, RG):
    AV = 1 + 50 * 10**3 / RG
    return ( (H1 - H0) * KH * S * AV )

#-- Métodos comunes a todos los ejercicios --
def valor_mas_probable (mediciones):
    return np.mean(mediciones, dtype=np.float64)

def incertidumbre_tipo_A (mediciones):
    s = np.std(mediciones, ddof=1, )     # ddof=1 -> grados de libertad v = 1/n-1
    return s/np.sqrt(len(mediciones))

def incertidumbre_tipo_B (error_relativo, error_cuentas, cuentas_indicadas, mediciones):
    return (error_relativo/100 + error_cuentas/cuentas_indicadas) * np.mean(mediciones) / np.sqrt(3)

def incertidumbre_tipo_B (error_relativo, error_cuentas, cuentas_indicadas, valor_mas_probable):
    return (error_relativo/100 + error_cuentas/cuentas_indicadas) * valor_mas_probable / np.sqrt(3)

def incertidumbre_combinada (uA, uB):
    return np.sqrt(uA**2 + uB**2)

<img src="Ejercicio_1.png" style="margin-left:auto; margin-right:auto"/>

***
#####

#### <u>Cálculo de valor más probable</u>

#####

Se obtienen los valores más probables de $R_1$ y $R_2$ a través de la siguiente expresión:

#####

\begin{equation*}
\overline{R_i} = \frac{1}{n} \cdot \sum\limits_{k=1}^{n} R_{i,k}
\end{equation*}

#####

Siendo:

> n: cantidad de mediciones

> $R_{i_{k}}$: i-ésimos valores de resistencias, k-ésimos valores de las n mediciones.

#####

Reemplazando los valores de la tabla, obtenemos $\overline{R_1}$ y $\overline{R_2}$:

#####

In [158]:
R1 = [2.01, 2.04, 2.03, 2.02, 2.04]
R2 = [0.53, 0.54, 0.52, 0.51, 0.53]

ov_R1 = valor_mas_probable(R1)
ov_R2 = valor_mas_probable(R2) 

display(Markdown("##\n" + f"$$\\overline{{R_1}} = {ov_R1} \, [\Omega]$$\n"))
display(Markdown("##\n" + f"$$\\overline{{R_2}} = {ov_R2} \, [\Omega]$$\n"))

##
$$\overline{R_1} = 2.028 \, [\Omega]$$


##
$$\overline{R_2} = 0.526 \, [\Omega]$$


#####

Se obtiene el valor más probable de $e_m$ a través de la siguiente expresión:

#####

\begin{equation*}
\overline{e_m} = \frac{ \overline{ R_2 } }{ \overline{ R_2 } - \overline{ R_1 } }
\end{equation*}

#####

In [159]:
ov_em = calculo_em (ov_R1, ov_R2)

display(Markdown("##\n" + f"$$\\overline{{e_m}} = {ov_em}$$\n"))

##
$$\overline{e_m} = -0.35019973368841545$$


#####

#### <u>Cálculo de incertidumbres</u>

#####

<u>Incertidumbre tipo A:</u>

#####

Se obtienen las incertidumbres tipo A a través de la siguiente expresión:

#####

\begin{equation*}
u_A(R_i) = \frac{ s(R_i) }{ \sqrt{n} } = \frac{ \sqrt{ \frac{ 1 }{ n-1 } \sum \limits_{k=1}^{n} \left( R_{i,k} - \overline{R_i} \right)^{2} } }{ \sqrt{n} }
\end{equation*}

#####

In [153]:
uA_R1 = incertidumbre_tipo_A (R1)
uA_R2 = incertidumbre_tipo_A (R2)

display(Markdown("##\n" + f"$$u_A(R_1) = {uA_R1} \, [\Omega]$$\n"))
display(Markdown("##\n" + f"$$u_A(R_2) = {uA_R2} \, [\Omega]$$\n"))

##
$$u_A(R_1) = 0.005830951894845336 \, [\Omega]$$


##
$$u_A(R_2) = 0.00509901951359279 \, [\Omega]$$


#####

<u>Incertidumbre tipo B:</u>

#####

Se obtienen las incertidumbres tipo B a través de la siguiente expresión:

#####

\begin{equation*}
u_B(R_i) = \frac{ \Delta R_i }{ \sqrt{3} } = \frac{ \left( \frac{er\%}{100} + \frac{n}{ci} \right) \cdot \overline{R_i} }{ \sqrt{3} } 
\end{equation*}

#####


Siendo:

> $er%$: Error relativo.

> $n$: Error de cuentas.

> $ci$: Cuentas indicadas.

#####

Además:

> $ci = \frac{Ind}{Res} \frac{ [u] }{ \left[ \frac{u}{cuenta} \right] }= [cuentas]$

> $Ind$: Indicación del instrumento con su respectiva $u$ unidad.

> $Res$: Resolución del instrumento con su respectiva $u$ unidad por cuenta, dada por el rango seleccionado y la cantidad de dígitos que el mismo puede representar.

#####


In [154]:
er = 0.5
n = 2

ind_R1 = ov_R1
res_R1 = 0.01
ci_R1 = ind_R1 / res_R1

ind_R2 = ov_R2
res_R2 = 0.01
ci_R2 = ind_R2 / res_R2

#def incertidumbre_tipo_B (error_relativo, error_cuentas, cuentas_indicadas, valor_mas_probable):
uB_R1 = incertidumbre_tipo_B (er, n, ci_R1, ov_R1)
uB_R2 = incertidumbre_tipo_B (er, n, ci_R2, ov_R2)

display(Markdown("##\n" + f"$$u_B(R_1) = {uB_R1} \, [\Omega]$$\n"))
display(Markdown("##\n" + f"$$u_B(R_2) = {uB_R2} \, [\Omega]$$\n"))

##
$$u_B(R_1) = 0.017401337113375322 \, [\Omega]$$


##
$$u_B(R_2) = 0.013065436591761232 \, [\Omega]$$


#####

<u>Incertidumbre de $R_i$:</u>

#####

Se obtienen las incertidumbres de $R_i$ a través de la siguiente expresión:

#####

\begin{equation*}
u(R_i) = \sqrt{ u_A^2(R_i) + u_B^2(R_i) }
\end{equation*}

#####

In [155]:
#def incertidumbre_combinada (uA, uB):
uC_R1 = incertidumbre_combinada (uA_R1, uB_R1)
uC_R2 = incertidumbre_combinada (uA_R2, uB_R2)

display(Markdown("##\n" + f"$$u(R_1) = {uC_R1} \, [\Omega]$$\n"))
display(Markdown("##\n" + f"$$u(R_2) = {uC_R2} \, [\Omega]$$\n"))

##
$$u(R_1) = 0.018352289593762786 \, [\Omega]$$


##
$$u(R_2) = 0.01402517854907143 \, [\Omega]$$


#####

<u> Incertidumbre combinada </u>

#####

Se obtienen la incertidumbre combinada de $e_m$ a través de la siguiente expresión:

#####

\begin{equation*} 
u_C(e_m) = \sqrt{\left[\frac{de_m}{dR_1}\right]^2 \cdot u^2(R_1) + \left[\frac{de_m}{dR_2}\right]^2 \cdot u^2(R_2) }
\end{equation*}

#####

\begin{equation*} 
u_C(e_m) = \sqrt{\left[ \frac{ \overline{R_2} }{ ( \overline{R_2} - \overline{R_1} )^2 } \right]^2 \cdot u^2(R_1) + \left[ \frac{ -\overline{R_1} }{ ( \overline{R_2} - \overline{R_1} )^2 } \right]^2 \cdot u^2(R_2) }
\end{equation*}

#####

In [156]:
dem_dR1 = ov_R2 / (ov_R2 - ov_R1)**2
dem_dR2 = -ov_R1 / (ov_R2 - ov_R1)**2

uC_em = np.sqrt( (dem_dR1)**2 * uC_R1**2 + (dem_dR2)**2 * uC_R2**2)

display(Markdown("##\n" + f"$$u_C(e_m) = {uC_em}$$\n"))

##
$$u_C(e_m) = 0.013314047982818046$$


#####

<u>Incertidumbre expandida:</u>

#####

\begin{equation*}
U(e_m) = u_C(e_m) \cdot k_{95\%}
\end{equation*}

#####

In [157]:
k_95 = 2
U_em = uC_em * k_95

display(Markdown("##\n" + f"$$U(e_m) = {U_em}$$\n" + "##"))

##
$$U(e_m) = 0.026628095965636093$$
##

#####

<u>Resultado de la medición:</u>

#####

\begin{equation*}
e_m = \overline{e_m} \pm U(e_m) 
\end{equation*}

#####

\begin{equation*}
e_m = \left( -0.35019973368841545 \pm 0.026628095965636093 \right) \qquad k = 2
\end{equation*}

#####

\begin{equation*}
\boxed{ e_m = \left( -0.350 \pm 0.027 \right) \qquad k = 2 }
\end{equation*}

#####

***